### DISCUSSION TITLE
Feature: Obtaining Geographic Coordinates from Google Maps URL in OsmAnd

### DISCUSSION THREAD

**Discussion by pricewatermouse** [osmandapp/OsmAnd#17365] [Ideas]
As a user who frequently receives or comes across Google Maps links shared by others, I want to be able to open these links directly in OsmAnd so that I can easily explore the area or navigate to a specific point. Also I want to be able to share particular points from Google Maps application with OsmAnd application.

1. When I am viewing a point of interest on Google Maps, I can press the "Share" button to generate a shareable URL. At this time the generated Google Maps URL contains no geographic coordinates.
2. In OsmAnd, when I paste the Google Maps URL, the application should **automatically query the URL to retrieve the geographic coordinates**. The curl examples follow.
3. OsmAnd should be able to extract the geographic coordinates from the HTTP response.
4. The extracted geographic coordinates should be used to display the corresponding location on the map in OsmAnd.

### 1) Google Maps URL generated by Web version
```
curl -I https://goo.gl/maps/jnrT3N5z4yaPHh7y7
HTTP/2 302 
...
location: https://www.google.com/maps/place/Brandenburg+Gate/@52.5162746,13.3777041,17z/data=!3m1!4b1!4m6!3m5!1s0x47a851c655f20989:0x26bbfb4e84674c63!8m2!3d52.5162746!4d13.3777041!16zL20vMDE0a2Y4?entry=tts&shorturl=1
content-length: 0
...
```
_Extract coordinates 52.5162746,13.3777041 from the location header_

### 2) Google Maps URL generated by Android application
```
curl -I https://maps.app.goo.gl/PJdcARfqULFnpENj7
HTTP/2 302 
...
location: https://www.google.com/maps/place/Brandenburg+Gate,+Pariser+Platz,+10117+Berlin,+Germany/data=!4m2!3m1!1s0x47a851c655f20989:0x26bbfb4e84674c63?utm_source=mstt_1&entry=gps&lucs=47068609&g_ep=CAESCjExLjgxLjE1MDQYACDXggMqCDQ3MDY4NjA5QgJMSw%3D%3D
content-length: 0
...
```
_The coordinates are missing. Follow the redirect to find ones:_

```
curl -L https://maps.app.goo.gl/PJdcARfqULFnpENj7
HTTP/2 200
<!DOCTYPE html><html itemscope="" 
...
[[[1,191]],3,null,0,197]],null,\"Pariser Platz, 10117 Berlin, Germany\",null,null,\"https://www.google.com/maps/preview/place/Brandenburg+Gate,+Pariser+Platz,+10117+Berlin,+Germany/@52.5162746,13.3777041,2428a,13.1y/data\\u003d!4m2!3m1!1s0x47a851c655f20989:0x26bbfb4e84674c63\",
...
```
_Use regular expression like @(-?\d+\.\d+),(-?\d+\.\d+) to extract geographic coordinates 52.5162746,13.3777041 from response body_

## Note:
Consider caching of recent [Google Maps URL => coordinates] pairs for offline usage.

----

![p1](https://github.com/osmandapp/OsmAnd/assets/47009757/490e138a-a540-43c7-abe6-40baae0183ea)

![p2](https://github.com/osmandapp/OsmAnd/assets/47009757/feb6dd96-d5b8-4264-aaa3-7605b6d13ff3)

**Comment by ThomasMeister123** [comment]
I would also like this....

**Comment by vshcherb** [comment]
With approach you showed it could be implemented. So it's worth to create an issue or pull request ;-)

**Comment by floppy-engineer** [comment]
Maybe it would be easier to implement it as a plug-in. Sorry, it's just a suggestion. I haven't checked the possibility by analyzing the plug-in API.

**Comment by vshcherb** [comment]
I think someone could file a complain to EU committee about monopoly situation here. There are hundreds formats of coordinates and OLC created by Google but it intentionally uses super hidden links to not allow share location from Google Maps
